gemma-rest/src/main/java/ubic/gemma/rest/DatasetsWebService.java

```diff
@@ -1186,6 +1190,32 @@ public Response getDatasetRawExpression( @PathParam("dataset") DatasetArg<?> dat
                 .build();
     }
 
+    @GZIP
+    @GET
+    @Path("/{dataset}/data/singleCell")
+    @Produces(MediaTypeUtils.TEXT_TAB_SEPARATED_VALUES_UTF8)
+    @Operation(summary = "Retrieve single-cell expression data of a dataset")
+    public Response getDatasetSingleCellExpression( @PathParam("dataset") DatasetArg<?> datasetArg,
+            @QueryParam("quantitationType") QuantitationTypeArg<?> quantitationTypeArg ) {
+        ExpressionExperiment ee = datasetArgService.getEntity( datasetArg );
+        QuantitationType qt;
+        if ( quantitationTypeArg != null ) {
+            qt = quantitationTypeArgService.getEntity( quantitationTypeArg, ee, SingleCellExpressionDataVector.class );
+        } else {
+            qt = singleCellExpressionExperimentService.getPreferredSingleCellQuantitationType( ee )
+                    .orElseThrow( () -> new NotFoundException( "No preferred single-cell quantitation type could be found for " + ee + "." ) );
+        }
+        java.nio.file.Path p;
+        try {
+            p = expressionDataFileService.writeOrLocateTabularSingleCellExpressionData( ee, qt );
```

Comment:
If the file is not present, that can be quite costly to perform.

Also, we might be serving an incomplete fie in the meantime.